### DISCUSSION TITLE
Any way to use line/column numbers in matchxpath?

### DISCUSSION THREAD

**Discussion by Machine-Maker** [checkstyle/checkstyle#16075] [Q&A] [answered]
I'd like to have an xpath query that checks if 2 tokens are on the same line

**Comment by romani** [comment]
We do not them in model, at time we implemented it we thought it is too advanced and might never be required.
If you share your idea on validation we might agree that it is valid reason to introduce such  attributes in xpath.

**Reply by Machine-Maker** [reply]
What I was trying to do, is make sure that specific annotations are always on the same lines as a method. Basically trying to make sure that method return type annotations are always on the same line as the method return type (which is(should) always be on the same line as the method ident token)

EDIT: I'm not super knowledgeable about xpath and what you can and cannot do with it, so I don't even know if it would be possible to construct such a query that would fail even if the line numbers were exposed as attributes.

**Reply by romani** [reply, marked as answer]
You can try to use https://checkstyle.org/checks/annotation/annotationonsameline.html

And suppress by xpath or any other filter all violations expect for names you need.

You need to use negative look forward in regexp (Negative Lookahead)

**Reply by Machine-Maker** [reply]
Ok, so I can do that, and yeah it seems to work. 

But side note, does property expansion not apply to the suppression files? I'd love to store a list of TYPE_USE annotations somewhere else and just pipe it into the suppression filter regular expression but I can't seem to get it to work.

**Reply by romani** [reply]
Please share config as CLI example https://checkstyle.org/report_issue.html#How_to_report_a_bug.3F
To let us see all details, there are numerous nuances, without exact config and exact java file , it hard to say how to make it work.

**Reply by romani** [reply]
Without details from you, I can suggest to look at https://checkstyle.org/filters/suppressionsinglefilter.html this module is representation of single suppression tag in file. It helps when people don't like more than single config for checkstyle. Main config for sure has support of variables.

**Reply by romani** [reply]
@Machine-Maker, it would be awesome if you share your configs to let others see a solution.

**Reply by romani** [reply]
@Machine-Maker , 

>does property expansion not apply to the suppression files?

Please report issue on it, and share use case of usage to let us that it is not artificial demand. We like real user stories and use cases.

**Reply by Machine-Maker** [reply]
Yeah, no problem. 
```xml
    <module name="SuppressionSingleFilter">
        <!--Suppresses warnings except for common TYPE_USE that should be on same line-->
        <property name="checks" value="AnnotationOnSameLine"/>
        <property name="message" value="^Annotation '(?!(${type_use_annotations}))"/>
    </module>
```
and 
```xml
        <module name="AnnotationOnSameLine">
            <!--matches all annotations, but most are suppressed in above-->
            <property name="tokens" value="METHOD_DEF"/>
        </module>
```
Here is my solution, and `${type_user_annotations}` is just a property that is populate from the gradle extension with a list of all the type_use_annotations (e.g. `(NonNull|NotNull|Nullable)`. In the future I probably will implement some simple classpath scanning to automatically accumulate all the annotations that are marked as such.
